File: training-data/python/merge_two_sorted_lists.py
class ListNode:
    def __init__(self, val, nxt=None):
        self.val, self.next = val, nxt


def from_list(xs):
    head = None
    for x in reversed(xs):
        head = ListNode(x, head)
    return head


def to_list(node):
    out = []
    while node:
        out.append(node.val)
        node = node.next
    return out


def merge(a, b):
    dummy = tail = ListNode(0)
    while a and b:
        if a.val <= b.val:
            tail.next, a = a, a.next
        else:
            tail.next, b = b, b.next
        tail = tail.next
    tail.next = a or b
    return dummy.next


if __name__ == "__main__":
    print(to_list(merge(from_list([1, 2, 4, 9]), from_list([1, 3, 4, 5, 10]))))

File: training-data/perl/merge_two_sorted_lists.pl
use strict;
use warnings;

sub from_list { my $head; $head = { val => $_, next => $head } for reverse @_; $head }
sub to_list   { my ($n, @out) = @_; for (; $n; $n = $n->{next}) { push @out, $n->{val} } @out }

sub merge {
    my ($a, $b) = @_;
    my $dummy = {};
    my $tail = $dummy;
    while ($a && $b) {
        if ($a->{val} <= $b->{val}) { $tail->{next} = $a; $a = $a->{next} }
        else                        { $tail->{next} = $b; $b = $b->{next} }
        $tail = $tail->{next};
    }
    $tail->{next} = $a || $b;
    return $dummy->{next};
}

print join(' ', to_list(merge(from_list(1, 2, 4, 9), from_list(1, 3, 4, 5, 10)))), "\n";
print join(' ', to_list(merge(undef, from_list(7)))), "\n";
